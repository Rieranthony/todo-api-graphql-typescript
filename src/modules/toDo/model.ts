import * as mongoose from "mongoose";
import { IToDo, ToDoStatus } from "../../interfaces/toDo.interface";

const toDoSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true
  },
  description: String,
  status: {
    type: String,
    enum: Object.values(ToDoStatus),
    default: ToDoStatus.PENDING,
    required: true
  },
  deletedAt: Date
}).set("timestamps", true);

const ToDoModel = mongoose.model<IToDo & mongoose.Document>("ToDo", toDoSchema);

export default class Model {
  public create = async (data: Partial<IToDo>): Promise<IToDo> => {
    const report = new ToDoModel(data);
    return await report.save();
  };

  public updateBy = async (
    key: keyof IToDo,
    keyValue: any,
    data: Partial<IToDo>
  ): Promise<Partial<IToDo> | null> => {
    const report = await ToDoModel.findOneAndUpdate({ [key]: keyValue }, data, {
      new: true,
      runValidators: true,
      upsert: true
    })
      .lean()
      .exec();

    return report;
  };

  public delete = async (_id: string): Promise<Partial<IToDo> | null> => {
    return await this.updateBy("_id", _id, {
      deletedAt: new Date()
    });
  };

  public findBy = async (
    key: keyof IToDo,
    value: any
  ): Promise<Partial<IToDo>> => {
    const result = await ToDoModel.find({ [key]: value })
      .limit(1)
      .lean()
      .exec();

    return result[0];
  };

  public getAll = async (): Promise<Partial<IToDo>[]> => {
    const result = await ToDoModel.find({ deletedAt: null }).lean();

    return result;
  };
}
