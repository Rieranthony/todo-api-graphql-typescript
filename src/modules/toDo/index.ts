export { default as Model } from "./model";
export { default as Service } from "./service";
export { default as TypeDef } from "./typeDef";
export { default as Resolver } from "./resolver";
