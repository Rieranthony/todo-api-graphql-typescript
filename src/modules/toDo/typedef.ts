import { gql } from "apollo-server-express";

export default gql`
  type ToDo {
    _id: ID!
    title: String!
    description: String
    status: ToDoStatus
    dueDate: Date!
    createdAt: Date!
    updatedAt: Date!
    deletedAt: Date
  }

  input ToDoInput {
    _id: ID
    title: String
    description: String
    status: ToDoStatus
    dueDate: Date
    deletedAt: Date
  }

  extend type Query {
    getAllToDo: [ToDo]!
    getToDo(_id: ID!): ToDo
  }

  extend type Mutation {
    createToDo(input: ToDoInput): ToDo!
    updateToDo(input: ToDoInput): ToDo
    deleteToDo(_id: ID!): ToDo
  }
`;
