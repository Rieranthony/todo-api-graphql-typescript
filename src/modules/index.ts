import { makeExecutableSchema } from "graphql-tools";
import { gql } from "apollo-server-express";
import * as _ from "lodash";
import { GraphQLSchema } from "graphql";

import * as ToDo from "./toDo";

import { initService } from "../utils";

/*
 * This is due to the Apollo limitation, so we can
 * extend query and Mutation in every typeDefs
 */
const rootDef = gql`
  scalar Date

  enum ToDoStatus {
    PENDING
    DONE
  }

  type Query {
    root: String
  }

  type Mutation {
    root: String
  }
`;

export const initSchemaFromModules = (): GraphQLSchema => {
  // Init services
  const services = {
    toDo: initService(ToDo)
  };

  // Merge all the resolver
  const mergedResolvers = _.merge(ToDo.Resolver(services.toDo));

  // Return the final graphql schema
  return makeExecutableSchema({
    typeDefs: [rootDef, ToDo.TypeDef],
    resolvers: {
      ...mergedResolvers
    }
  });
};
