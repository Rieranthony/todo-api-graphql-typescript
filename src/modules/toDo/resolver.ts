import { ApolloError } from "apollo-server-express";

import ToDoService from "./service";
import { IToDo, IToDoInput, ToDoStatus } from "../../interfaces/toDo.interface";
import { ResolverMap } from "../../interfaces/common.interfaces";

import { ERROR_ID_MISSING, ERROR_ARGS_INVALID } from "../../constants";
import logger from "../../logger";

export default (toDoService: ToDoService): ResolverMap<IToDo> => {
  return {
    Query: {
      getAllToDo: async () => {
        logger.info("Get all to do requested.");

        return await toDoService.getAll();
      },
      getToDo: async (_, { _id }: { _id: string }) => {
        logger.info("Get to do with _id: ", _id);

        return await toDoService.findBy("_id", _id);
      }
    },
    Mutation: {
      createToDo: async (_, args: { input: IToDoInput }) => {
        // TODO: check inputs using Yup for instance
        const { _id, status, ...data } = args.input;

        if (_id) {
          throw new ApolloError(
            "You can't create a to do with an ID",
            ERROR_ARGS_INVALID
          );
        }

        if (status === ToDoStatus.DONE) {
          throw new ApolloError(
            "You can't create a to do with a DONE status",
            ERROR_ARGS_INVALID
          );
        }

        logger.info("Delete to do with data: ", data);

        return await toDoService.create(data);
      },
      updateToDo: async (_, args: { input: IToDoInput }) => {
        // TODO: check inputs using Yup for instance
        const { _id, ...data } = args.input;

        if (!_id) {
          throw new ApolloError(
            "_id is missing to update the to do",
            ERROR_ID_MISSING
          );
        }

        logger.info("Delete to do with _id: ", _id, data);

        return await toDoService.updateBy("_id", _id, data);
      },
      deleteToDo: async (_, { _id }: { _id: string }) => {
        if (!_id) {
          throw new ApolloError(
            "_id is missing to delete the to do",
            ERROR_ID_MISSING
          );
        }

        logger.info("Delete to do with _id: ", _id);

        return await toDoService.delete(_id);
      }
    }
  };
};
