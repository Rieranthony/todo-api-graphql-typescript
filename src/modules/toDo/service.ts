import ToDoModel from "./model";
import { IToDo } from "../../interfaces/toDo.interface";

export default class ToDoService {
  private model: ToDoModel;

  constructor(model: ToDoModel) {
    this.model = model;
  }

  public create = async (data: Partial<IToDo>): Promise<IToDo> =>
    this.model.create(data);

  public getAll = async (): Promise<Partial<IToDo>[]> => this.model.getAll();

  public findBy = async (
    key: keyof IToDo,
    value: any
  ): Promise<Partial<IToDo>> => this.model.findBy(key, value);

  public delete = async (_id: string): Promise<Partial<IToDo> | null> =>
    this.model.delete(_id);

  public updateBy = async (
    key: keyof IToDo,
    keyValue: any,
    data: Partial<IToDo>
  ): Promise<Partial<IToDo> | null> => this.model.updateBy(key, keyValue, data);
}
