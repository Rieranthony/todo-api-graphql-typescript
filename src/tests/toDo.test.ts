import { createTestClient } from "apollo-server-testing";
import { ApolloServer, gql } from "apollo-server-express";

import { initSchemaFromModules } from "../modules";
import {
  connect,
  clearDatabase,
  closeDatabase,
  populateDatabase
} from "./utils";
import { IToDo, ToDoStatus } from "../interfaces/toDo.interface";
import { ToDoMongooseModel } from "../modules/toDo/model";

beforeAll(async () => connect());

afterEach(async () => {
  await clearDatabase();
});

afterAll(async done => {
  await closeDatabase();
  done();
});

const mockTodo: Partial<IToDo>[] = [
  {
    _id: "5e16075abba89142f8f2afd2",
    title: "Todo 1",
    description: "Desc 1",
    status: ToDoStatus.PENDING
  },
  {
    _id: "5e1607772c8db465edc44a4c",
    title: "Todo 2",
    description: "Desc 2",
    status: ToDoStatus.PENDING
  },
  {
    _id: "5e1607819c9eb8b8be169cf2",
    title: "Todo 3",
    description: "Desc 3",
    status: ToDoStatus.PENDING
  },
  {
    _id: "5e1607878d043e453fcdfea2",
    title: "Todo 4",
    description: "Desc 4",
    status: ToDoStatus.PENDING
  }
];

/**
 * ToDo test suite.
 */
describe("ToDo", () => {
  const graphQLSchema = initSchemaFromModules();

  it(`should return an empty array if there is no todo`, async () => {
    const server = new ApolloServer({
      schema: graphQLSchema
    });

    // use the test server to create a query function
    const { query } = createTestClient(server);

    const GET_ALL_TODO = gql`
      query getAllToDo {
        getAllToDo {
          _id
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await query({
      query: GET_ALL_TODO
    });

    expect(res).toMatchSnapshot();
  });

  it(`should return an array of to do`, async () => {
    const server = new ApolloServer({
      schema: graphQLSchema
    });

    // Populate database with todos
    await populateDatabase(ToDoMongooseModel, mockTodo);

    // use the test server to create a query function
    const { query } = createTestClient(server);

    const GET_ALL_TODO = gql`
      query getAllToDo {
        getAllToDo {
          _id
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await query({
      query: GET_ALL_TODO
    });

    expect(res).toMatchSnapshot();
  });

  it(`should return a specific todo`, async () => {
    const server = new ApolloServer({
      schema: graphQLSchema
    });

    const GET_TODO = gql`
      query getToDo($_id: ID!) {
        getToDo(_id: $_id) {
          _id
          status
        }
      }
    `;

    // use the test server to create a query function
    const { query } = createTestClient(server);

    // Populate database with todos
    await populateDatabase(ToDoMongooseModel, mockTodo);

    // run query against the server and snapshot the output
    const res = await query({
      query: GET_TODO,
      variables: {
        _id: mockTodo[1]._id
      }
    });

    expect(res).toMatchSnapshot();
  });

  it(`should create a to do`, async () => {
    const server = new ApolloServer({
      schema: graphQLSchema
    });

    // use the test server to create a query function
    const { mutate } = createTestClient(server);

    const CREATE_TO_DO = gql`
      mutation createToDo($input: ToDoInput!) {
        createToDo(input: $input) {
          title
          description
          status
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await mutate({
      mutation: CREATE_TO_DO,
      variables: {
        input: {
          title: "Test create todo",
          description: "Test create description todo"
        }
      }
    });

    expect(res).toMatchSnapshot();
  });

  it(`should update a to do`, async () => {
    const server = new ApolloServer({
      schema: graphQLSchema
    });

    // Populate database with todos
    await populateDatabase(ToDoMongooseModel, mockTodo);

    // use the test server to create a query function
    const { mutate } = createTestClient(server);

    const UPDATE_TO_DO = gql`
      mutation updateToDo($input: ToDoInput!) {
        updateToDo(input: $input) {
          _id
          title
          description
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await mutate({
      mutation: UPDATE_TO_DO,
      variables: {
        input: {
          id: mockTodo[0]._id,
          title: "Updated todo title",
          description: "Updated todo description",
          status: ToDoStatus.DONE
        }
      }
    });

    expect(res).toMatchSnapshot();
  });

  it(`should delete a to do`, async () => {
    const server = new ApolloServer({
      schema: graphQLSchema
    });

    // Populate database with todos
    await populateDatabase(ToDoMongooseModel, mockTodo);

    // use the test server to create a query function
    const { mutate } = createTestClient(server);

    const DELETE_TO_DO = gql`
      mutation deleteToDo($_id: ID!) {
        deleteToDo(_id: $_id) {
          _id
          title
          description
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await mutate({
      mutation: DELETE_TO_DO,
      variables: {
        id: mockTodo[0]._id
      }
    });

    expect(res).toMatchSnapshot();
  });
});
