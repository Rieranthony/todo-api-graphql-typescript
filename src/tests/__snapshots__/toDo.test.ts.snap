// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo should create a to do 1`] = `
Object {
  "data": Object {
    "createToDo": Object {
      "description": "Test create description todo",
      "status": "PENDING",
      "title": "Test create todo",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo should delete a to do 1`] = `
Object {
  "data": Object {
    "deleteToDo": Object {
      "_id": "5e16075abba89142f8f2afd2",
      "description": "Desc 1",
      "title": "Todo 1",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo should return a specific todo 1`] = `
Object {
  "data": Object {
    "getToDo": Object {
      "_id": "5e1607772c8db465edc44a4c",
      "status": "PENDING",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo should return an array of to do 1`] = `
Object {
  "data": Object {
    "getAllToDo": Array [
      Object {
        "_id": "5e16075abba89142f8f2afd2",
      },
      Object {
        "_id": "5e1607772c8db465edc44a4c",
      },
      Object {
        "_id": "5e1607819c9eb8b8be169cf2",
      },
      Object {
        "_id": "5e1607878d043e453fcdfea2",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo should return an empty array if there is no todo 1`] = `
Object {
  "data": Object {
    "getAllToDo": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo should update a to do 1`] = `
Object {
  "data": Object {
    "updateToDo": Object {
      "_id": "5e16075abba89142f8f2afd2",
      "description": "Updated todo description",
      "title": "Updated todo title",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo shouldn't create a to do with a DONE status in params 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You can't create a to do with a DONE status],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`ToDo shouldn't create a to do with a ID in params 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: You can't create a to do with an ID],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
